groups/csa/complexity_check.cpp

```diff
@@ -0,0 +1,143 @@
+};
+
+
+// Matches every if,switch,while,for,try,catch inside a method declaration
+static const internal::DynTypedMatcher& dyn_matchComplexity() 
+{
+    static const internal::DynTypedMatcher& matcher =
+    findAll(functionDecl(hasDescendant(stmt(eachOf(
+        findAll(stmt(ifStmt().bind("if"))), 
+        findAll(stmt(switchCase().bind("switch"))),
+        findAll(stmt(whileStmt().bind("while"))),
+        findAll(stmt(forStmt().bind("for"))),
+        findAll(stmt(tryStmt().bind("try"))),
+        findAll(stmt(catchStmt().bind("catch"))),
+        findAll(stmt(conditionalOperator().bind("?operator"))),
+        findAll(stmt(ifStmt(hasCondition(findAll(binaryOperator(hasOperatorName("&&")).bind("condOp")))))),
+        findAll(stmt(ifStmt(hasCondition(findAll(binaryOperator(hasOperatorName("||")).bind("condOp"))))))
+        )))).bind("methodDecl"));
+
+    return matcher; 
+}
+
+
+void report::operator()()
+{
+    MatchFinder finder;
+    csabase::OnMatch<report, &report::calculateComplex> m(this);
+    finder.addDynamicMatcher(dyn_matchComplexity(), &m);
+    finder.match(*d_analyser.context()->getTranslationUnitDecl(), *d_analyser.context());
+    generateReport();
+}
+
+// Generates report if any of the methods exceeded the verified threshold
+void report::generateReport(){
+    // Get the complexity threshold
+    data& d = a.attachment<data>();
+    std::string tag = "CX01";
+    std::string message = "Complexity Exceeded the maximum threshold ! Threshold verified: " 
+    + std::to_string(d.d_max_complexity) + " Found: ";
+    //location of method
+    SourceLocation loc; 
+    
+    for(auto curMethod : complexDict){
+        if (curMethod.second > d.d_max_complexity){
+            loc = curMethod.first->getLocStart();
+            std::string val = std::to_string(curMethod.second);
+            a.report(loc, check_name, tag, message + val);
+        }
+        // Add to total File complexity
+        complexity += curMethod.second;
+    }
+    
+}
+
+
+void report::calculateComplex(BoundNodes const & nodes)
+{
+    FunctionDecl const* curMethod = nodes.getNodeAs<FunctionDecl>("methodDecl");
+    IfStmt const * curIf = nodes.getNodeAs<IfStmt>("if");
+    SwitchCase const * curSwitch = nodes.getNodeAs<SwitchCase>("switch");
+    WhileStmt const * curWhile = nodes.getNodeAs<WhileStmt>("while");
+    ForStmt const * curFor = nodes.getNodeAs<ForStmt>("for");
+    CXXTryStmt const * curTry = nodes.getNodeAs<CXXTryStmt>("try");
+    CXXCatchStmt const * curCatch = nodes.getNodeAs<CXXCatchStmt>("catch");
+    BinaryOperator const * curBin = nodes.getNodeAs<BinaryOperator>("condOp");
+    ConditionalOperator const * curCond = nodes.getNodeAs<ConditionalOperator>("?operator");
+
+    SourceManager &sm = a.manager();
+    const auto& fId = sm.getMainFileID();
+    const auto& fileId = sm.getFileID(curMethod->getLocation());
+
+    // If matched method is not in the same .cpp file i.e:includes
+    if (fId == fileId){
```

Comment:
The analyser has functionality to do this in a cleaner way, just pass the FunctionDecl which is being processed and it will tell you if you can continue. Something like:

  ```  clang::Decl const * decl = nodes.getNodeAs<clang::Decl>("declaration");

    if (!a.is_component(decl)) {
        return;                                                       // RETURN
    }